ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java

```diff
@@ -138,8 +141,32 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
       environmentContext.putToProperties(HiveMetaHook.ALTER_TABLE_OPERATION_TYPE, desc.getType().name());
       if (partitions == null) {
         long writeId = desc.getWriteId() != null ? desc.getWriteId() : 0;
-        context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true,
-            writeId);
+        try {
+          context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true, writeId);
+        } catch (HiveException ex) {
+          if (environmentContext.getProperties().containsKey(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER) && Boolean
+              .valueOf(environmentContext.getProperties().get(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER))) {
+            // in case of rollback of alter table do the following:
+            // 1. drop the already altered table but keep the data files
+            // 2. recreate the original table
+            // 3. run msck repair to sync partitions on filesystem with metastore
```

Comment:
This could have unexpected result, if we have extra directories in the table root, which are not added yet to the HMS. I think it would be good to have a way to disable things

Reply:
I did some manual testing, with adding/removing directories to the table, and it seems `Msck#repair` is using the file system as a single source of truth. If some partitions are added/removed it updates the metastore respectively.

Reply:
That's what I fear. We might end up adding partitions which are on the FS but intentionally not has not been added to the table yet. Also I seem to recall that we had issues with `MSCK REPAIR` if the file naming was not correct (wrong order of partition keys, or wrong type of values - not valid number for an int column etc...)

Reply:
I've removed the drop/create/msck part, and rolled back the serde, input/output format, partition keys and table properties. An exception is also thrown to notice the user we had an issue during migration and a `msck repair table` should be issued to add the missing partition data to metastore.